src/utils/helpers/structure/generateProgress.ts

```diff
@@ -0,0 +1,82 @@
+import { TemplateElementCategory } from '../../generated/graphql'
+import { ElementStateNEW, FullStructure, PageNEW, Progress } from '../../types'
+
+const initialProgress = {
+  doneNonRequired: 0,
+  doneRequired: 0,
+  completed: false,
+  totalNonRequired: 0,
+  totalRequired: 0,
+  totalSum: 0,
+  valid: true,
+}
+
+const calculateCompleted = (totalSum: number, doneRequired: number, doneNonRequired: number) => {
+  const totalDone = doneRequired + doneNonRequired
+  return totalSum === totalDone
+}
+
+const getSectionProgress = (pages: PageNEW[]): Progress => {
+  return pages.reduce((sectionProgress: Progress, { progress }) => {
+    if (!progress) return sectionProgress
+    let {
+      doneNonRequired,
+      doneRequired,
+      completed,
+      totalNonRequired,
+      totalRequired,
+      totalSum,
+      valid,
+    } = sectionProgress
+    doneNonRequired += progress.doneNonRequired || 0
+    totalNonRequired += progress.totalNonRequired || 0
+    doneRequired += progress.doneRequired || 0
+    totalRequired += progress.totalRequired || 0
+    totalSum = totalRequired + doneNonRequired
+    completed = calculateCompleted(totalSum, doneRequired, doneNonRequired)
+    if (!progress.valid) valid = false
+    return {
+      doneNonRequired,
+      doneRequired,
+      completed,
+      totalNonRequired,
+      totalRequired,
+      totalSum,
+      valid,
+    }
+  }, initialProgress)
+}
+
+export const generateResponsesProgress = (structure: FullStructure) => {
+  Object.values(structure.sections).forEach((section) => {
+    Object.values(section.pages).forEach((page) => {
+      const progress: Progress = initialProgress
```

Comment:
does this need to be spread ? I think it would store it as reference otherwise ? (and all pages will have the same progress ?)

Reply:
Yes. This is happening! Oh, haven't noticed on this PR but on the following one.